### PR TITLE
Add SpecialMobius CoordinateMap

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

The current BinaryCompactObject Domain (#709) has its compact objects located at the centers of two cubes. By construction, these objects can neither be translated nor scaled along the x-axis - they are frozen in place at the center of each cube. One problem that this causes is that the center of mass is not necessarily located at the origin of domain. SpecialMobius is a conformal map with analytic inverse that can be used to shift the center of mass to the origin while still preserving the spherical shapes of the wave zone and the two compact objects.

The name comes from the fact that it is a special case of the more general Mobius family of conformal transformations. Rename suggestions are welcome.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — COMMENTED** [review]
Determine a good limit for mu and enforce it

**Comment by mar-celine** [comment]
@kidder Changes have been addressed

**Review by kidder — COMMENTED** [review]
squash

File: src/Domain/CoordinateMaps/SpecialMobius.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  const double mu_squared = mu * mu;
+  const ReturnType r_squared = square(x) + square(y) + square(z);
+  const ReturnType common_factor =
+      (mu_squared - 1.0) / square(1.0 - 2.0 * mu * x + mu_squared * r_squared);
+  auto jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(coords[0]), 0.0);
+
+  get<0, 0>(jacobian_matrix) =
+      -(mu_squared * (2.0 * x * x - r_squared) - 2.0 * mu * x + 1.0) *
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.cpp]: square(x)

File: src/Domain/CoordinateMaps/SpecialMobius.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      common_factor;
+  get<1, 0>(jacobian_matrix) = (2.0 * mu * y * (mu * x - 1.0)) * common_factor;
+  get<2, 0>(jacobian_matrix) = (2.0 * mu * z * (mu * x - 1.0)) * common_factor;
+
+  get<0, 1>(jacobian_matrix) = -1.0 * get<1, 0>(jacobian_matrix);
+  get<1, 1>(jacobian_matrix) =
+      -(mu_squared * (r_squared - 2.0 * y * y) - 2.0 * mu * x + 1.0) *
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.cpp]: square(y)

File: src/Domain/CoordinateMaps/SpecialMobius.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      common_factor;
+  get<2, 1>(jacobian_matrix) = 2.0 * mu_squared * y * z * common_factor;
+
+  get<0, 2>(jacobian_matrix) = -1.0 * get<2, 0>(jacobian_matrix);
+  get<1, 2>(jacobian_matrix) = get<2, 1>(jacobian_matrix);
+  get<2, 2>(jacobian_matrix) =
+      -(mu_squared * (r_squared - 2.0 * z * z) - 2.0 * mu * x + 1.0) *
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.cpp]: square(z)

File: src/Domain/CoordinateMaps/SpecialMobius.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>
+SpecialMobius::mobius_distortion_jacobian(const std::array<T, 3>& coords,
+                                          const double mu) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& x = coords[0];
+  const ReturnType& y = coords[1];
+  const ReturnType& z = coords[2];
+  const double mu_squared = mu * mu;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.cpp]: square(mu)

File: src/Domain/CoordinateMaps/SpecialMobius.hpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * \brief Redistributes gridpoints within the unit sphere.
+ * \image html SpecialMobius.png "A sphere with a `mu` of 0.25."
+ *
+ * \details A special case of the conformal Mobius transformation that
+ * maps the unit ball to itself. This map depends on a single
+ * parameter, `mu` \f$ = \mu\f$, which is the x-coordinate of the preimage
+ * of the origin under this map. This map has the fixed points \f$x=1\f$ and
+ * \f$x=-1\f$.
+ *
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.hpp]: mention limit of mu (for accuracy)

File: src/Domain/CoordinateMaps/SpecialMobius.hpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ * takes the form \f$ f(z) = \frac{az+b}{cz+d}\f$, where
+ * \f$z, a, b, c, d \in \mathbb{C}\f$, and \f$ad-bc\neq 0\f$.
+ *
+ * The special case used in this map is the function
+ * \f$ f(z) = \frac{z - \mu}{1 - z\mu}\f$. This has the desired properties:
+ * - The unit disk in the complex plane is mapped to itself.
+ *
+ * - The x-axis is mapped to itself.
+ *
+ * - \f$(\mu,0) = 0\f$.
```
Comment (kidder) [inline src/Domain/CoordinateMaps/SpecialMobius.hpp]: fix this

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
@osheamonn could you look at this one? I have the dox rendered: http://gh.nilsdeppe.com/spectre/pr_798/classCoordinateMaps_1_1SpecialMobius.html

**Review by nilsdeppe — APPROVED** [review]

**Review by osheamonn — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

**Review by wthrowe — APPROVED** [review]

**Comment by wthrowe** [comment]
Travis failures are the usual timeouts.

**Pull request merged**